In [1]:
import pandas as pd
import os

path = "/workspaces/codespaces-jupyter/data/YAAD/"
stimulus_descripttion_file = pd.read_excel(path + 'Stimulus_Description.xlsx')

fear_video_id = stimulus_descripttion_file.query('`Target Emotion` == "fear"').head(1).get('Video ID').item()

path_single_modal_data = path + "RawData/Singlemodal/ECG/"
path_multi_modal_data = path + "RawData/Multimodal/ECG/"

single_modal_data_files = os.listdir(path_single_modal_data)
multi_modal_data_files = os.listdir(path_multi_modal_data)

single_modal_dfs = []
for file in single_modal_data_files:
    if "v" + str(fear_video_id) in file:
        df = pd.read_csv(path_single_modal_data + file, delimiter='\t')
        single_modal_dfs.append(df)

multi_modal_dfs = []
for file in multi_modal_data_files:
    if "v" + str(fear_video_id) in file:
        df = pd.read_csv(path_multi_modal_data + file, delimiter='\t')
        multi_modal_dfs.append(df)

In [2]:
def str_dfs_to_real_dfs(dfs) -> pd.array:
    real_dfs = []
    for df in dfs:
        arr = df.columns.values
        str_vals = arr[0].split(',')
        real_vals = [float(each) for each in str_vals]
        real_vals_pd = pd.array(real_vals)
        real_dfs.append(real_vals_pd)
    return pd.array(real_dfs)

def mean_of_means(dfs) -> float:
    t = []
    for each in dfs:
        t.append(each.mean())
    return pd.array(t).mean()

def min_of_mins(dfs) -> float:
    t = []
    for each in dfs:
        t.append(each.min())
    return pd.array(t).min()

def max_of_maxs(dfs) -> float:
    t = []
    for each in dfs:
        t.append(each.max())
    return pd.array(t).max()

def mean_of_vars(dfs) -> float:
    t = []
    for each in dfs:
        t.append(each.var())
    return pd.array(t).mean()

In [3]:
real_sm_dfs = str_dfs_to_real_dfs(single_modal_dfs)
real_mm_dfs = str_dfs_to_real_dfs(multi_modal_dfs)

single refers to singlemodal data
multi refers to multimodal data

In [4]:
print("mean of means single: ", mean_of_means(real_sm_dfs))
print("mean of means multi: ", mean_of_means(real_mm_dfs))

print("min of mins single: ", min_of_mins(real_sm_dfs))
print("min of mins multi: ", min_of_mins(real_mm_dfs))

print("max of maxs single: ", max_of_maxs(real_sm_dfs))
print("max of maxs multi: ", max_of_maxs(real_mm_dfs))

print("mean of vars single: ", mean_of_vars(real_sm_dfs))
print("mean of vars multi: ", mean_of_vars(real_mm_dfs))

mean of means single:  -18.522971550372002
mean of means multi:  -10.029048553400726
min of mins single:  -297.21
min of mins multi:  -21.106
max of maxs single:  18.678
max of maxs multi:  8.1532
mean of vars single:  0.13076550723324526
mean of vars multi:  0.041651835869248255


In [5]:
single_modal_df = []
columns = list(range(1, 5001)) + ["Vtype"]

# We put vtype 1 for fear sessions, and vtype 0 for everything else
for file in single_modal_data_files:
    df = pd.read_csv(path_single_modal_data + file, delimiter='\t')
    vtype = float(file.split('v')[1][0])

    vtype = 1 if vtype == 7.0 else 0

    rdf = str_dfs_to_real_dfs([df])[0]
    minval = rdf.min()
    rdf_and_vtype = pd.array([(each - minval) for each in rdf] + [vtype])
    single_modal_df.append(rdf_and_vtype)
single_modal_df = pd.DataFrame(single_modal_df, columns=columns)


multi_modal_df = []
for file in multi_modal_data_files:
    df = pd.read_csv(path_multi_modal_data + file, delimiter='\t')
    vtype = float(file.split('v')[1][0])

    vtype = 1 if vtype == 7.0 else 0

    rdf = str_dfs_to_real_dfs([df])[0]
    minval = rdf.min()
    rdf_and_vtype = pd.array([(each - minval) for each in rdf] + [vtype])
    multi_modal_df.append(rdf_and_vtype)
multi_modal_df = pd.DataFrame(multi_modal_df, columns=columns)


In [6]:
single_y = single_modal_df.Vtype
single_X = single_modal_df[list(range(1, 5001))]

multi_y = multi_modal_df.Vtype
multi_X = multi_modal_df[list(range(1, 5001))]

First, we train using single modal data and validate 

In [7]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import mean_absolute_error

fear_detect_model = DecisionTreeClassifier(random_state=1)

fear_detect_model.fit(single_X, single_y)

val_predictions = fear_detect_model.predict(multi_X)
print(mean_absolute_error(multi_y, val_predictions))

0.11507936507936507


Next, we train using multi modal data (inverse)

In [8]:
fear_detect_model.fit(multi_X, multi_y)

val_predictions = fear_detect_model.predict(single_X)
print(mean_absolute_error(single_y, val_predictions))

0.12337662337662338


training for single shows a bit less mean error
Now, we will try to find optimal max_leaf_nodes 

In [9]:
def try_leaves(max_leaf_nodes, train_X, train_y, val_X, val_y):
    fear_detect_model = DecisionTreeClassifier(max_leaf_nodes=max_leaf_nodes, random_state=1)
    fear_detect_model.fit(train_X, train_y)

    val_predictions = fear_detect_model.predict(val_X)
    return mean_absolute_error(val_y, val_predictions)

In [10]:
for leaves in (10, 50, 100, 500, 1000, 5000):
    print(f"max_leaf_nodes = {leaves}, train-single mae: ", try_leaves(leaves, single_X, single_y, multi_X, multi_y))
    print(f"max_leaf_nodes = {leaves}, train-multi mae: ", try_leaves(leaves, multi_X, multi_y, single_X, single_y))

max_leaf_nodes = 10, train-single mae:  0.11507936507936507
max_leaf_nodes = 10, train-multi mae:  0.12987012987012986


max_leaf_nodes = 50, train-single mae:  0.11507936507936507
max_leaf_nodes = 50, train-multi mae:  0.12337662337662338
max_leaf_nodes = 100, train-single mae:  0.11507936507936507
max_leaf_nodes = 100, train-multi mae:  0.12337662337662338
max_leaf_nodes = 500, train-single mae:  0.11507936507936507
max_leaf_nodes = 500, train-multi mae:  0.12337662337662338
max_leaf_nodes = 1000, train-single mae:  0.11507936507936507
max_leaf_nodes = 1000, train-multi mae:  0.12337662337662338
max_leaf_nodes = 5000, train-single mae:  0.11507936507936507
max_leaf_nodes = 5000, train-multi mae:  0.12337662337662338


max_leaf_node=50 with singlemodal data as training data seems to be optimal setting for current configuration.

In [18]:
fear_detect_model = DecisionTreeClassifier(max_leaf_nodes=50, random_state=1)

fear_detect_model.fit(single_X, single_y)

# try to test our model on fear sessions
fear_multi_modal =  multi_modal_df.query("`Vtype` == 1")
print(fear_multi_modal.info())
val_X = fear_multi_modal[list(range(1, 5001))]
print(val_X.info())
val_predictions = fear_detect_model.predict(val_X)
print(val_predictions)
# as all of the entries in val_prediction should ideally be equal to 7.0:
counter = 0
for each in val_predictions:
    if round(each, 0) == 1.0:
        counter += 1
print(counter)
# getting the % of correct answers:
print(100* (counter / len(val_predictions)))


<class 'pandas.core.frame.DataFrame'>
Index: 36 entries, 0 to 248
Columns: 5001 entries, 1 to Vtype
dtypes: float64(5001)
memory usage: 1.4 MB
None
<class 'pandas.core.frame.DataFrame'>
Index: 36 entries, 0 to 248
Columns: 5000 entries, 1 to 5000
dtypes: float64(5000)
memory usage: 1.4 MB
None
[0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 1. 1. 1. 1. 0. 1. 0. 0. 0. 1. 0. 0. 1. 0.
 0. 0. 1. 1. 0. 0. 0. 0. 1. 1. 1. 1.]
17
47.22222222222222


accuracy is 47%
TT

Lets try processing the data

In [43]:
import numpy as np
import pandas as pd
import neurokit2 as nk
import matplotlib.pyplot as plt

# Generate synthetic signals
# ecg = nk.ecg_simulate(duration=10, heart_rate=70)

ecg_data = real_sm_dfs[0]

# Load raw ECG data (replace with your data)
# Assuming 'ecg_data' is a list or array of 5000 data points
# ecg_data = [your_raw_ecg_data]  # Replace with actual data
sampling_rate = 5000 / 39  # Approximate sampling rate

# Step 1: Clean the ECG signal
ecg_cleaned = nk.ecg_clean(ecg_data, sampling_rate=sampling_rate)
print(ecg_cleaned)
# Step 2: Process the ECG signal to find R-peaks
signals, info = nk.ecg_process(ecg_cleaned, sampling_rate=sampling_rate)
# nk.ecg_plot(signals, info)
print(signals["ECG_Clean"].values)

# Step 3: Extract R-peaks, indexes where R-peaks are located
r_peaks = info["ECG_R_Peaks"] # np.ndarray
tl = r_peaks.tolist()
print(type(tl[0]))

# Step 4: Delineate QRS complex
delineate = nk.ecg_delineate(ecg_cleaned, r_peaks, sampling_rate=sampling_rate, method="peak")

# Access the Q, R, and S points
q_points = delineate[0]["ECG_Q_Peaks"]
s_points = delineate[0]["ECG_S_Peaks"]

# Step 5: Plot the ECG with identified R-peaks, Q, and S points
# plt.plot(ecg_cleaned, label="ECG Signal")
# plt.scatter(r_peaks, ecg_cleaned[r_peaks], color="red", label="R-peaks")
# plt.scatter(q_points, ecg_cleaned[q_points], color="blue", label="Q points")
# plt.scatter(s_points, ecg_cleaned[s_points], color="green", label="S points")
# plt.legend()
# plt.show()

# nk.hrv(r_peaks, sampling_rate=sampling_rate, show=True)



[ 0.00351687 -0.0001571  -0.01133531 ... -0.00811949 -0.00942752
 -0.00815506]
[-0.00634845 -0.01122435 -0.01985871 ... -0.00122269 -0.00088196
  0.00080364]
<class 'int'>


In [46]:
# we will repeat the previous approach, only just add the ecg_process method, and remove normalisation
cleaned_single_modal_df = []
columns = list(range(1, 5001)) + ["Vtype"]
sampling_rate = 5000 / 39

# We put vtype 1 for fear sessions, and vtype 0 for everything else
for file in single_modal_data_files:
    df = pd.read_csv(path_single_modal_data + file, delimiter='\t')
    vtype = float(file.split('v')[1][0])

    vtype = 1 if vtype == 7.0 else 0

    rdf = str_dfs_to_real_dfs([df])[0]
    rdf_cleaned, info = nk.ecg_process(rdf, sampling_rate=sampling_rate)
    rdf_and_vtype = pd.array([each for each in rdf_cleaned["ECG_Clean"].values[info["ECG_R_Peaks"]]] + [vtype])
    cleaned_single_modal_df.append(rdf_and_vtype)
cleaned_single_modal_df = pd.DataFrame(single_modal_df, columns=columns)


cleaned_multi_modal_df = []
for file in multi_modal_data_files:
    df = pd.read_csv(path_multi_modal_data + file, delimiter='\t')
    vtype = float(file.split('v')[1][0])

    vtype = 1 if vtype == 7.0 else 0

    rdf = str_dfs_to_real_dfs([df])[0]
    rdf_cleaned, info = nk.ecg_process(rdf, sampling_rate=sampling_rate)
    rdf_and_vtype = pd.array([each for each in rdf_cleaned["ECG_Clean"].values[info["ECG_R_Peaks"]]] + [vtype])
    cleaned_multi_modal_df.append(rdf_and_vtype)
cleaned_multi_modal_df = pd.DataFrame(multi_modal_df, columns=columns)

In [47]:
cleaned_single_y = cleaned_single_modal_df.Vtype
cleaned_single_X = cleaned_single_modal_df[list(range(1, 5001))]

cleaned_multi_y = cleaned_multi_modal_df.Vtype
cleaned_multi_X = cleaned_multi_modal_df[list(range(1, 5001))]

In [56]:
max_leaf_nodes = 50
cleaned_fear_detect_model = DecisionTreeClassifier(max_leaf_nodes=max_leaf_nodes, random_state=1)

cleaned_fear_detect_model.fit(cleaned_single_X, cleaned_single_y)

cleaned_val_predictions = cleaned_fear_detect_model.predict(cleaned_multi_X)
print(mean_absolute_error(cleaned_multi_y, cleaned_val_predictions))

0.11507936507936507


In [57]:
# try to test our model on fear sessions
cleaned_fear_multi_modal =  cleaned_multi_modal_df.query("`Vtype` == 1")
print(cleaned_fear_multi_modal.info())
cleaned_val_X = cleaned_fear_multi_modal[list(range(1, 5001))]
print(cleaned_val_X.info())
cleaned_val_predictions = cleaned_fear_detect_model.predict(cleaned_val_X)
print(cleaned_val_predictions)
# as all of the entries in val_prediction should ideally be equal to 1.0:
counter = 0
for each in cleaned_val_predictions:
    if round(each, 0) == 1.0:
        counter += 1
print(counter)
# getting the % of correct answers:
print(100* (counter / len(cleaned_val_predictions)))

<class 'pandas.core.frame.DataFrame'>
Index: 36 entries, 0 to 248
Columns: 5001 entries, 1 to Vtype
dtypes: float64(5001)
memory usage: 1.4 MB
None
<class 'pandas.core.frame.DataFrame'>
Index: 36 entries, 0 to 248
Columns: 5000 entries, 1 to 5000
dtypes: float64(5000)
memory usage: 1.4 MB
None
[0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 1. 1. 1. 1. 0. 1. 0. 0. 0. 1. 0. 0. 1. 0.
 0. 0. 1. 1. 0. 0. 0. 0. 1. 1. 1. 1.]
17
47.22222222222222


Anyway, we got the same result. Now, lets try to give more data for the training, and less for predicting. Divide traing to test as 75:25.

In [ ]:
from sklearn.model_selection import train_test_split

# Combine the datasets
X = pd.concat([cleaned_single_X, cleaned_multi_X], ignore_index=True)
y = pd.concat([cleaned_single_y, cleaned_multi_y], ignore_index=True)

# Split the data into training and validation sets
train_X, test_val_X, train_y, test_val_y = train_test_split(X, y, test_size=0.25, random_state=1)

# Initialize and train the model
bigger_fear_detect_model = DecisionTreeClassifier(max_leaf_nodes=max_leaf_nodes, random_state=1)
bigger_fear_detect_model.fit(train_X, train_y)

# Make predictions and evaluate the model
bigger_val_predictions = bigger_fear_detect_model.predict(test_val_X)
print(mean_absolute_error(test_val_y, bigger_val_predictions))

0.12745098039215685


In [90]:
# try to test our model on fear sessions
test_fear =  pd.DataFrame(test_val_y).query("`Vtype` == 1")
fear_val_X = X[X.index.isin(test_fear.index.values)]
fear_val_predictions = bigger_fear_detect_model.predict(fear_val_X)

# as all of the entries in val_prediction should ideally be equal to 1.0:
counter = 0
for each in fear_val_predictions:
    if round(each, 0) == 1.0:
        counter += 1
print(counter, len(fear_val_predictions))
# getting the % of correct answers:
print(100* (counter / len(fear_val_predictions)))

8 14
57.14285714285714


The accuracy is now 57%, and have been increased by ~10%. Great!